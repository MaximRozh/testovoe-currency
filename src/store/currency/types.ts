import CurrencyModel from "../../models/Currency.model";

export interface CurrencyState {
  currencies: CurrencyModel[];
  selectedCur: string;
  currenciesCode: string[];
}

export enum CurrencyActionEnum {
  GET_CURRENCY = "GET_CURRENCY",
  SELECT_CURRENCY = "SELECT_CURRENCY",
  SET_CURRENCY_CODE = "SET_CURRENCY_CODE",
}

export interface GetCurrencyAction {
  type: CurrencyActionEnum.GET_CURRENCY;
  payload: CurrencyModel[];
}

export interface SelectCurrencyAction {
  type: CurrencyActionEnum.SELECT_CURRENCY;
  payload: string;
}

export interface SetCurrenciesCodeAction {
  type: CurrencyActionEnum.SET_CURRENCY_CODE;
  payload: string[];
}

export type CurrencyAction =
  | GetCurrencyAction
  | SelectCurrencyAction
  | SetCurrenciesCodeAction;
