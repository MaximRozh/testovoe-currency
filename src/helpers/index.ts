import CurrencyModel from "../models/Currency.model";

export const BASE_CURRENCY = "UAH";
export const BTC_CURRENCY = "BTC";

interface SplitedText {
  amount: number;
  fromCurrency: string;
  toCurrency: string;
}
export const splitedText = (str: string): SplitedText => {
  ///[in]/ig use regEx
  const splited = str.split(" ").filter((item) => item !== "in");
  return {
    amount: +splited[0],
    fromCurrency: splited[1],
    toCurrency: splited[2],
  };
};

export const checkInputText = (
  fromCurrency: string,
  toCurrency: string,
  amount: number
): boolean => {
  const isNotEmpty = fromCurrency && toCurrency && amount;
  const isNotEqual = fromCurrency !== toCurrency;
  const hasCurrencyCode =
    fromCurrency?.length === 3 && toCurrency?.length === 3;
  return amount > 0 && Boolean(isNotEmpty) && isNotEqual && hasCurrencyCode;
};

export const isHaveCurrencyInArray = (
  values: string,
  arr: CurrencyModel[]
): boolean => {
  return arr.some((item) => item.ccy === values?.toUpperCase());
};

export const getCurrencyIndex = (arr: any[], value: string): number =>
  arr.findIndex((item) => item.ccy === value?.toUpperCase());

export const isUah = (val: string): boolean =>
  val?.toUpperCase() === BASE_CURRENCY;

export const isBtc = (val: string): boolean =>
  val?.toUpperCase() === BTC_CURRENCY;

export const toFixed = (val: number): number => Number(val.toFixed(3));

export const validation = (
  fromCurrency: string,
  toCurrency: string,
  amount: number,
  currencies: CurrencyModel[]
) => {
  const isValid = checkInputText(fromCurrency, toCurrency, amount);
  const validCurrencyFrom = isHaveCurrencyInArray(fromCurrency, currencies);
  const validCurrencyTo = isHaveCurrencyInArray(toCurrency, currencies);
  return (
    !isValid ||
    !(
      (validCurrencyFrom || isUah(fromCurrency)) &&
      (validCurrencyTo || isUah(toCurrency))
    )
  );
};
