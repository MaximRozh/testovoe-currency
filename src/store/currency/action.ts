import {
  CurrencyActionEnum,
  GetCurrencyAction,
  SelectCurrencyAction,
  SetCurrenciesCodeAction,
} from "./types";

import CurrencyModel from "../../models/Currency.model";
import { AppDispatch } from "..";
import CurrencyService from "../../services/currencyService";

export const CurrencyActionCreator = {
  setCurencies: (data: CurrencyModel[]): GetCurrencyAction => ({
    type: CurrencyActionEnum.GET_CURRENCY,
    payload: data,
  }),
  setCurrenciesCode: (data: string[]): SetCurrenciesCodeAction => ({
    type: CurrencyActionEnum.SET_CURRENCY_CODE,
    payload: data,
  }),
  //try to fix ANY
  getCurencies:
    (): any =>
    (dispatch: AppDispatch): void => {
      CurrencyService.getCurrency().then((res) => {
        dispatch(CurrencyActionCreator.setCurencies(res));

        const currenciesCode = res.map((item: CurrencyModel) => item.ccy);
        dispatch(
          CurrencyActionCreator.setCurrenciesCode(["UAH", ...currenciesCode])
        );
      });
    },
  selectCurrency: (currency: string): SelectCurrencyAction => ({
    type: CurrencyActionEnum.SELECT_CURRENCY,
    payload: currency,
  }),
};
