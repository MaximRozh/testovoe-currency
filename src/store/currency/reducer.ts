import { CurrencyState, CurrencyActionEnum, CurrencyAction } from "./types";

const initialState: CurrencyState = {
  currencies: [],
  selectedCur: "UAH",
  currenciesCode: [],
};

export default function currencyReducer(
  state = initialState,
  action: CurrencyAction
): CurrencyState {
  switch (action.type) {
    case CurrencyActionEnum.GET_CURRENCY:
      return { ...state, currencies: action.payload };
    case CurrencyActionEnum.SELECT_CURRENCY:
      return { ...state, selectedCur: action.payload };
    case CurrencyActionEnum.SET_CURRENCY_CODE:
      return { ...state, currenciesCode: action.payload };
    default:
      return state;
  }
}
