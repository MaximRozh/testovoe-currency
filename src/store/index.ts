import { applyMiddleware, combineReducers, createStore } from "redux";
import thunk from "redux-thunk";
import currencyReducer from "./currency/reducer";

const rootReducer = combineReducers({
  currencyReducer,
});

export const store = createStore(rootReducer, applyMiddleware(thunk));

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
