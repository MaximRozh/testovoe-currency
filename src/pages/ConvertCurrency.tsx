import React, { useEffect, useState } from "react";
import { useDispatch } from "react-redux";
import ConvertContent from "../components/ConvertContent";
import HeroImage from "../components/HeroImage";
import Layout from "../components/Layout";
import {
  BASE_CURRENCY,
  getCurrencyIndex,
  isBtc,
  isUah,
  splitedText,
  toFixed,
  validation,
} from "../helpers";
import { useDebaunce } from "../hooks/useDebaunce";
import { useTypedSelector } from "../hooks/useTypedSelector";
import { CurrencyActionCreator } from "../store/currency/action";

const ConvertCurrency = () => {
  const { currencies, selectedCur } = useTypedSelector(
    (state) => state.currencyReducer
  );

  const dispatch = useDispatch();

  const [value, setValue] = useState<string>("");
  const [result, setResult] = useState<string>("");
  const [hasError, setHasError] = useState<boolean>(false);

  useEffect(() => {
    if (!currencies.length) {
      dispatch(CurrencyActionCreator.getCurencies());
    }
  }, [dispatch, currencies]);

  const convertTextIntoResult = (text: string) => {
    const { fromCurrency, toCurrency, amount } = splitedText(text);
    const validate = validation(fromCurrency, toCurrency, amount, currencies);
    //validation
    if (validate) {
      setHasError(true);
      setResult("");
      return;
    }
    setHasError(false);

    //find rates index
    const indexFrom = getCurrencyIndex(currencies, fromCurrency);
    const indexTo = getCurrencyIndex(currencies, toCurrency);

    const rateForm = currencies[indexFrom]?.buy || 1;
    const rateTo = currencies[indexTo]?.buy || 1;

    //calculation
    //for UAH
    let rate = 0;

    const fromCurrencyIsUAH = isUah(fromCurrency);
    const toCurrencyIsUAH = isUah(toCurrency);

    if (fromCurrencyIsUAH || toCurrencyIsUAH) {
      rate = fromCurrencyIsUAH ? amount / rateTo : amount * rateForm;
    }

    //for BTC
    const fromCurrencyIsBTC = isBtc(fromCurrency);
    const toCurrencyIsBTC = isBtc(toCurrency);

    if (fromCurrencyIsBTC || toCurrencyIsBTC) {
      //cause bts is only in usd currency
      const usdRate = currencies.find((item) => item.ccy === "USD")!.buy;

      rate = fromCurrencyIsBTC
        ? ((rateForm * usdRate) / rateTo) * amount
        : (rateForm / usdRate / rateTo) * amount;
    }

    const hasBtc = fromCurrencyIsBTC || toCurrencyIsBTC;
    const hasUah = fromCurrencyIsUAH || toCurrencyIsUAH;

    //result
    const sum = hasUah || hasBtc ? rate : (amount * rateForm) / rateTo;
    const resultValuete = toCurrencyIsUAH
      ? BASE_CURRENCY
      : currencies[indexTo]?.ccy;

    setResult(`${toFixed(sum)} ${resultValuete}`);
  };

  const convert = useDebaunce(convertTextIntoResult, 500);

  const handler = (value: string) => {
    setValue(value);
    convert(value);
  };

  return (
    <Layout>
      <div className="currency-wrapper">
        <ConvertContent
          value={value}
          handler={handler}
          result={result}
          hasError={hasError}
          selectedCur={selectedCur}
        />
        <HeroImage />
      </div>
    </Layout>
  );
};

export default ConvertCurrency;
