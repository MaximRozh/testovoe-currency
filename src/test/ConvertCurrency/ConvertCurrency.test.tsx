import React from "react";
import { cleanup, fireEvent, RenderResult } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ConvertCurrency from "../../pages/ConvertCurrency";
import { store } from "../../store";
import { CurrencyActionCreator } from "../../store/currency/action";
import { act } from "react-dom/test-utils";
import { toFixed } from "../../helpers";
import {
  currencies,
  ExchangeCurrency,
  getRatesFromCurrencies,
} from "../testUtils";
import { renderWithProviders } from "../renderWithProvider";

jest.useFakeTimers("modern");
let div: RenderResult;
let result: HTMLSpanElement;
let input: HTMLInputElement;

const amount = 15;

describe("check calculation", () => {
  beforeEach(() => {
    div = renderWithProviders(<ConvertCurrency />);
    act(() => {
      store.dispatch(CurrencyActionCreator.setCurencies(currencies));
    });
    result = div.container.querySelector(
      ".convert-result-text"
    ) as HTMLSpanElement;
    input = div.container.querySelector(
      "input[name=conver]"
    ) as HTMLInputElement;
  });
  afterEach(cleanup);

  it("Convert To UAH", () => {
    userEvent.type(
      input,
      `${amount} ${ExchangeCurrency.USD} in ${ExchangeCurrency.UAH}`
    );
    act(() => {
      jest.runAllTimers();
    });
    const { USD } = getRatesFromCurrencies([ExchangeCurrency.USD], currencies);

    const calc = toFixed(USD * amount);

    const resultValue: string = result?.innerHTML || "";

    expect(parseFloat(resultValue)).toBe(calc);
  });
  it("Convert From UAH", () => {
    userEvent.type(
      input,
      `${amount} ${ExchangeCurrency.UAH} in ${ExchangeCurrency.USD}`
    );
    act(() => {
      jest.runAllTimers();
    });
    const { USD } = getRatesFromCurrencies([ExchangeCurrency.USD], currencies);

    const calc = toFixed(amount / USD);

    const resultValue: string = result?.innerHTML || "";

    expect(parseFloat(resultValue)).toBe(calc);
  });
  it("Convert From BTC", () => {
    userEvent.type(
      input,
      `${amount} ${ExchangeCurrency.BTC} in ${ExchangeCurrency.EUR}`
    );
    act(() => {
      jest.runAllTimers();
    });

    const { USD, BTC, EUR } = getRatesFromCurrencies(
      [ExchangeCurrency.USD, ExchangeCurrency.BTC, ExchangeCurrency.EUR],
      currencies
    );

    const calc = toFixed(((BTC * USD) / EUR) * amount);

    const resultValue: string = result?.innerHTML || "";

    expect(parseFloat(resultValue)).toBe(calc);
  });
  it("Convert To BTC", () => {
    const amountForBTC = amount * 1000;

    userEvent.type(
      input,
      `${amountForBTC} ${ExchangeCurrency.EUR} in ${ExchangeCurrency.BTC}`
    );
    act(() => {
      jest.runAllTimers();
    });

    const { USD, BTC, EUR } = getRatesFromCurrencies(
      [ExchangeCurrency.USD, ExchangeCurrency.BTC, ExchangeCurrency.EUR],
      currencies
    );

    const calc = toFixed((EUR / USD / BTC) * amountForBTC);

    const resultValue: string = result?.innerHTML || "";

    expect(parseFloat(resultValue)).toBe(calc);
  });
  it("Has error", () => {
    userEvent.type(input, `${amount} WRONG_VALUE in ${ExchangeCurrency.BTC}`);
    act(() => {
      jest.runAllTimers();
    });
    fireEvent.focusOut(input);
    const errorText = div?.container.querySelector(".error-text");

    expect(errorText).toBeInTheDocument();
  });
});

describe("ConvertCurrency snapshot", () => {
  it("Make ConvertCurrency snapshot", () => {
    const div = renderWithProviders(<ConvertCurrency />);
    const { asFragment } = div;
    expect(asFragment()).toMatchSnapshot();
  });
});
