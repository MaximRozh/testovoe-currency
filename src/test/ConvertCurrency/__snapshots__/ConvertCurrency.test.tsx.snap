// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConvertCurrency snapshot Make ConvertCurrency snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="nav-wrapper"
    >
      <div
        class="nav-wrapper-container"
      >
        <nav
          class="nav-bar"
        >
          <a
            aria-current="page"
            class="active"
            href="/"
          >
            <img
              alt="Logo"
              class="logo"
              src="Label.svg"
            />
          </a>
          <ul
            class="nav-menu"
          >
            <li
              class="nav-menu-item"
            >
              <a
                aria-current="page"
                class="active-link"
                href="/"
              >
                Currency converter
              </a>
            </li>
            <li
              class="nav-menu-item"
            >
              <a
                class=""
                href="/current-currency"
              >
                Exchange rates
              </a>
            </li>
          </ul>
          <select />
        </nav>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="currency-wrapper"
      >
        <div
          class="currency-convert"
        >
          <div
            class="convert-content"
          >
            <h2
              class="convert-header"
            >
              ONLILE CURRENCY CONVERTER
            </h2>
            <p
              class="convert-text"
            >
              Fast, simplle, non-custodial Exchange Service. Check live foreign currency exchange rates
            </p>
            <input
              class="convert-input "
              name="conver"
              placeholder="15 usd in uah"
              value=""
            />
          </div>
          <div
            class="convert-result"
          >
            <span
              class="convert-result-text"
            >
              0 UAH
            </span>
          </div>
        </div>
        <div
          class="hero-image-container"
        >
          <img
            alt="Hero banner"
            class="hero-image"
            src="hero.svg"
            style="opacity: 0; transition: opacity 500ms ease-in 0ms;"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
